File: backend/bot.js
import dotenv from "dotenv";
import TelegramBot from "node-telegram-bot-api";

dotenv.config();

const token = process.env.TELEGRAM_BOT_TOKEN;
const miniAppUrl = process.env.MINI_APP_URL;
const menuButtonText = process.env.MENU_BUTTON_TEXT || "Open Real Holat";

if (!token) {
  console.error("Missing TELEGRAM_BOT_TOKEN in .env");
  process.exit(1);
}

if (!miniAppUrl) {
  console.error("Missing MINI_APP_URL in .env");
  process.exit(1);
}

const bot = new TelegramBot(token, { polling: true });

const webAppKeyboard = {
  reply_markup: {
    inline_keyboard: [
      [
        {
          text: menuButtonText,
          web_app: { url: miniAppUrl }
        }
      ]
    ]
  }
};

async function configureMenuButton() {
  try {
    await bot.setChatMenuButton({
      menu_button: {
        type: "web_app",
        text: menuButtonText,
        web_app: { url: miniAppUrl }
      }
    });
    console.log("Menu button configured.");
  } catch (error) {
    console.warn("Could not set menu button automatically. Configure in BotFather if needed.");
    console.warn(error.message);
  }
}

bot.onText(/\/start/, async (msg) => {
  const chatId = msg.chat.id;
  const fullName = [msg.from?.first_name, msg.from?.last_name].filter(Boolean).join(" ") || "Citizen";

  await bot.sendMessage(
    chatId,
    `Welcome, ${fullName}.\n\nReal Holat is a public transparency platform where reports are visible to everyone.\nTap below to open the app.`,
    webAppKeyboard
  );
});

bot.onText(/\/open/, async (msg) => {
  await bot.sendMessage(msg.chat.id, "Open Real Holat:", webAppKeyboard);
});

bot.on("polling_error", (error) => {
  console.error("Polling error:", error.message);
});

configureMenuButton().finally(() => {
  console.log("Real Holat bot backend is running.");
  console.log(`Mini App URL: ${miniAppUrl}`);
});

File: backend/bot.py
import logging
import os
from dataclasses import dataclass

from dotenv import load_dotenv
from telegram import InlineKeyboardButton, InlineKeyboardMarkup, MenuButtonWebApp, Update, WebAppInfo
from telegram.ext import Application, CommandHandler, ContextTypes


@dataclass
class BotConfig:
    token: str
    mini_app_url: str
    menu_button_text: str


load_dotenv()

logging.basicConfig(
    format="%(asctime)s | %(levelname)s | %(name)s | %(message)s",
    level=logging.INFO,
)
logger = logging.getLogger("real-holat-bot")


def read_config() -> BotConfig:
    token = os.getenv("TELEGRAM_BOT_TOKEN", "").strip()
    mini_app_url = os.getenv("MINI_APP_URL", "").strip()
    menu_button_text = os.getenv("MENU_BUTTON_TEXT", "Open Real Holat").strip() or "Open Real Holat"

    if not token:
        raise ValueError("Missing TELEGRAM_BOT_TOKEN in .env")
    if not mini_app_url:
        raise ValueError("Missing MINI_APP_URL in .env")
    if not mini_app_url.startswith("https://"):
        raise ValueError("MINI_APP_URL must start with https://")

    return BotConfig(token=token, mini_app_url=mini_app_url, menu_button_text=menu_button_text)


config = read_config()


def web_app_markup() -> InlineKeyboardMarkup:
    return InlineKeyboardMarkup(
        [[InlineKeyboardButton(text=config.menu_button_text, web_app=WebAppInfo(url=config.mini_app_url))]]
    )


async def start(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    user = update.effective_user
    full_name = " ".join([part for part in [user.first_name if user else None, user.last_name if user else None] if part])
    full_name = full_name or "Citizen"

    message = (
        f"Welcome, {full_name}.\n\n"
        "Real Holat is a public transparency platform where reports are visible to everyone.\n"
        "Tap below to open the app."
    )

    await update.effective_chat.send_message(message, reply_markup=web_app_markup())


async def open_app(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    await update.effective_chat.send_message("Open Real Holat:", reply_markup=web_app_markup())


async def set_menu_button(application: Application) -> None:
    try:
        await application.bot.set_chat_menu_button(
            menu_button=MenuButtonWebApp(text=config.menu_button_text, web_app=WebAppInfo(url=config.mini_app_url))
        )
        logger.info("Menu button configured.")
    except Exception as error:  # pragma: no cover
        logger.warning("Could not set menu button automatically. Configure in BotFather if needed.")
        logger.warning(str(error))


def main() -> None:
    application = Application.builder().token(config.token).build()

    application.add_handler(CommandHandler("start", start))
    application.add_handler(CommandHandler("open", open_app))

    application.post_init = set_menu_button

    logger.info("Real Holat Python bot backend is running.")
    logger.info("Mini App URL: %s", config.mini_app_url)
    application.run_polling(allowed_updates=Update.ALL_TYPES)


if __name__ == "__main__":
    main()
